### ISSUE TITLE
Change shutil.rmtree and os.walk to support very deep hierarchies

### ISSUE THREAD

**Issue by ad83914d-c8e2-4b82-8a39-fccff147f12c** [python/cpython#89727]
BPO | [45564](https://bugs.python.org/issue45564)
--- | :---

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-10-21.22:21:02.146>
labels = ['library', '3.9', 'type-crash']
title = 'shutil.rmtree and os.walk are implemented using recursion, fail on deep hierarchies'
updated_at = <Date 2021-10-21.22:21:02.146>
user = 'https://bugs.python.org/AlexanderPatrakov'
```

bugs.python.org fields:
```python
activity = <Date 2021-10-21.22:21:02.146>
actor = 'Alexander.Patrakov'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2021-10-21.22:21:02.146>
creator = 'Alexander.Patrakov'
dependencies = []
files = []
hgrepos = []
issue_num = 45564
keywords = []
message_count = 1.0
messages = ['404687']
nosy_count = 1.0
nosy_names = ['Alexander.Patrakov']
pr_nums = []
priority = 'normal'
resolution = None
stage = None
status = 'open'
superseder = None
type = 'crash'
url = 'https://bugs.python.org/issue45564'
versions = ['Python 3.9']
```

</p></details>

### Linked PRs
* gh-99803
* gh-100282
* gh-100347
* gh-100671
* gh-100703
* gh-103164
* gh-103234
* gh-103566
* gh-119634
* gh-119638
* gh-119748
* gh-119749
* gh-119764
* gh-119765
* gh-119766
* gh-119767
* gh-119768
* gh-119808
* gh-119918
* gh-119919

**Comment by barneygale** [comment]
This also affects `pathlib.Path.walk()`

**Comment by merwok** [comment]
I am unsure on the bug vs feature classification here.  On one side it seems like a clear problem and the fixes don’t change function signatures or the way they are used, but on the other side there is no guarantee that extremely deep hierachies are supported, and I wonder if there could be negative effects from the ocde changes (for example, can the `stack` list/deque grow extremely large during traversal?).

(Adding PR reviewers: @carljm @brettcannon @serhiy-storchaka)

**Comment by carljm** [comment]
I don't have strong feelings about calling it a bug vs a feature; I would tend to call it a bug because the function just fails in a subset of cases that aren't obviously outside its domain. I agree that it's a bug that we _could_ just document as a known limitation, though I don't see any reason to do that when the fix is not that difficult. I guess the only real impact of how it's classified is whether the fix would be backported?

> I wonder if there could be negative effects from the ocde changes (for example, can the stack list/deque grow extremely large during traversal?)

Sure, but in the current code the Python stack would instead grow very large in the same scenario. And the Python stack frames (plus everything they reference) are almost certainly larger than the stack elements tracked in the iterative version, so if anything I would expect the iterative version to also save memory in the case of a very deep traversal.

**Comment by merwok** [comment]
Yes exactly, the impact of my question is backporting or not.
Thanks for the reply about the memory concern!

I take it you feel ok about backporting the fix; let’s wait to see what a core dev thinks.

**Comment by brettcannon** [comment]
I would classify this as a feature request. There are plenty of things in CPython for which you can you run out of stack space for and it isn't considered a bug in those instances (and that's just part of general limitations that CPython has).

**Comment by carljm** [comment]
(To be clear, I don't think there's a strong case for backporting this, so reasoning backwards from the conclusion I'm quite happy to call it a feature :P )

**Comment by barneygale** [comment]
Are `glob.glob(recursive=True)` and `pathlib.Path.rglob()` also affected?

**Comment by jonburdo** [comment]
> Are `glob.glob(recursive=True)` and `pathlib.Path.rglob()` also affected?

Yes, it looks like it. `_rlistdir` in `Lib/os.py` and `_RecursiveWildcardSelector` in `Lib/pathlib.py` could be changed to fix it

**Comment by barneygale** [comment]
Some previous discussion in #70732.

**Comment by barneygale** [comment]
Am I right in thinking that `os.walk(followlinks=True)` will now consume all available memory if a symlink loop is encountered? Previously it would raise `RecursionError`.

**Comment by barneygale** [comment]
I'm not convinced we should implement this feature request.

A well-designed portable application shouldn't create very deep directory trees. For one thing, you'll hit `MAX_PATH` shenanigans on Windows long before you hit the Python recursion limit.

The remaining use case is where something has gone wrong, e.g. buggy code has created nonsense trees, or a malicious archive has been extracted. That seems sufficiently niche that calling `sys.setrecursionlimit()` before `shutil.rmtree()` is a reasonable solution.

**Comment by jonburdo** [comment]
> I'm not convinced we should implement this feature request.

I understand the hesitation, but I think it's worth it and have some thoughts in addition to comments in support in [this discussion thread](https://discuss.python.org/t/using-iterative-filesystem-walk-instead-of-recursive/21955)

> A well-designed portable application shouldn't create very deep directory trees. For one thing, you'll hit `MAX_PATH` shenanigans on Windows long before you hit the Python recursion limit.

Applications are not the only things that create directory trees. For many applications, portability is not a design goal, and many applications that are used are not well-designed. I'd also argue that your point highlights that `shutil.rmtree` is not currently portable itself. By default, it can delete basically any normal directory tree on some systems but not on others.

> The remaining use case is where something has gone wrong, e.g. buggy code has created nonsense trees, or a malicious archive has been extracted. That seems sufficiently niche that calling `sys.setrecursionlimit()` before `shutil.rmtree()` is a reasonable solution.

I think deep directory trees are less common, but I wouldn't say wanting to delete them when they occur is niche.

A few other thoughts:

1. In any situation where the call stack is already close to the recursion limit, you don't need a deep directory tree in order to hit the recursion limit. A function might recursively traverses a graph and then call `os.walk` or `shutil.rmtree`.

2. Recursion is an implementation detail which is not exposed to the user in the documentation for all of these functions, but limits the user in unexpected ways. For `shutil.rmtree` in particular, there's no implication that recursion is even used, and I was surprised to find that it was. Python users may easily have no idea that a recursion limit exists.
 
3. I think the speed improvements alone are significant enough to warrant these changes (over 6-7% in my `os.walk` benchmarks)

In my view the cost in terms of complexity is very low, and the benefit is significant. Any significant downsides I might be missing?

**Comment by JelleZijlstra** [comment]
> Am I right in thinking that `os.walk(followlinks=True)` can now consume all available memory if a symlink loop is encountered? Previously it would raise `RecursionError`.

This seems like an important concern, @Ovsyanka83 could you comment on this situation?

**Comment by Ovsyanka83** [comment]
I am currently gathering info on this issue. Will reply within ~2 hours with a short report on OS X, Windows, and Linux.

**Comment by Ovsyanka83** [comment]
I have created a symlink loop on 3 OSes and tested how they act. I am obviously no expert and my results are obviously not covering the entire playing field but they might help us understand just how bad the situation is.

On all three OSes an OSError gets raised and caught on the line of code attached.
![image](https://user-images.githubusercontent.com/17561586/210239004-492b3597-02fa-404c-aa50-6d43424cc17c.png)


1. On manjaro linux (kernel: 6.0.11-1, filesystem: ext4), I get an `OSError(40, 'Too many levels of symbolic links')` after walk yields 41 times.
2. On windows 10, I get an `OSError(22, 'The name of the file cannot be resolved by the system'` after walk yields 32 times. (I believe the main reason is the maximum path length and that we could get more yields if I started from `C:` but that does not affect our results).
3. On OS X Monterey, I get an `OSError(62, 'Too many levels of symbolic links')` after walk yields 33 times.

Seems like the OS is preventing us from shooting ourselves in the knee in every single case so I would not worry about these OSes too much. However, that only covers some platforms where python is used. What about others though?

**Comment by barneygale** [comment]
Thank you @jonburdo and @Ovsyanka83, you've put my mind at ease :-)

**Comment by pochmann** [comment]
I think you changed `os.walk`'s behavior. The recursive one supported late modifications of subdirs lists, due to being lazier. The iterative one doesn't (if I'm not mistaken... I can't test).

With topdown, you can modify a directory's subdirs, for example remove one and it won't be visited. Usually you'd do that while your walk is on that parent directory. But with the recursive one, it was possible to do it later. With the iterative one, I don't think that works anymore.

Here's a demo where I remove the second subdir after having walked onto the first:
```
import os

# Create demo dir with three subdirs
os.makedirs('demo/a')
os.makedirs('demo/b')
os.makedirs('demo/c')

# Walk onto "demo"
walk = os.walk('demo')
demo = next(walk)
print(demo)

# Walk onto first subdir
first = next(walk)
print(first)

# Remove the second subdir
del demo[1][1]

# Walk onto the remaining subdirs
for x in walk:
    print(x)
```
Output ([Try it online!](https://tio.run/##bZCxDoJADIb3e4pGB2HBGBdj4uQbuDgYh0OKXJArOarg02MLaqLhckPv69/@vdZPLsivN3Xoe1fVFBioMWYO@4CWETKsCDIXoHVcABcBEZp7KqQx1CSVLVHjaKHCpV3EEzSdpBeh4nO0txLIM8FM8cy0CnYyRaLRKBbpMMgOPHYcaSI2dXCeI@V/jXIx4feUZnxMFA6JofIgLR4on5Ov4YV89qnN8DYs4LQ6y/01UXXAyjrv/PW7kpwCdOA8qNPWgJzRrYv7/gU "Python 3.8 (pre-release) – Try It Online")), note that the second subdir wasn't walked:
```
('demo', ['c', 'a', 'b'], [])
('demo/c', [], [])
('demo/b', [], [])
```

[Excerpt from the recursive one](https://github.com/python/cpython/blob/a7a450f84a087421603170c2dad226bb881d4d9a/Lib/os.py#L412-L419):
```
        for dirname in dirs:
            new_path = join(top, dirname)
                yield from _walk(new_path, topdown, onerror, followlinks)
```
The iteration of `dirs` and the walking of the subdirs are intertwined. After walking a subdir, the paused iteration of `dirs` resumes. That allows the late modifications of `dirs` to have an effect.

[Excerpt from the iterative one](https://github.com/python/cpython/blob/73097d91a64620ae7f620705864b84234d85cc82/Lib/os.py#L418-L425):
```
            for dirname in reversed(dirs):
                new_path = join(top, dirname)
                    stack.append(new_path)
```
This eagerly puts the entire `dirs` contents onto the stack, before they're getting walked, and then `dirs` is never used again. So modifying `dirs` during the subdirs walking doesn't have an effect anymore.

**Comment by barneygale** [comment]
Is that used in practice? The documentation seems to imply that *dirnames* should only be modified before iteration resumes (my emphasis):

> When topdown is True, the caller can modify the dirnames list in-place (perhaps using [del](https://docs.python.org/3/reference/simple_stmts.html#del) or slice assignment), and [walk()](https://docs.python.org/3/library/os.html#os.walk) will only recurse into the subdirectories whose names remain in dirnames; this can be used to prune the search, impose a specific order of visiting, or even to inform [walk()](https://docs.python.org/3/library/os.html#os.walk) about directories the caller creates or renames **before it resumes [walk()](https://docs.python.org/3/library/os.html#os.walk) again**.

**Comment by pochmann** [comment]
I don't know whether it's used. But someone *might*, and I think justifiably, since:

<details><summary>arguments, click to see</summary>

- The way that doc part is written, I think the part you emphasized only applies to *"to inform walk() about directories the caller creates or renames"*.
- The first half of that doc part looks like the "specification" of the feature, and doesn't say the modifications need to be done before continuing.
- The second half of that doc part looks like *examples*, listing a few things the feature "can" be used for. It doesn't say that it *can't* be used for anything else. One other thing that's not listed but that I'd expect to work is to *duplicate* entries so they get walked multiple times.

</details>

That said, I agree it's unusual and wouldn't be surprised if nobody did that.

If we do want to restore the old behavior, I think we could instead put an iterator (forward, not reversed) on the stack:
```
            stack.append((
                new_path
                for dirname in dirs
                for new_path in [join(top, dirname)]
                if followlinks or not islink(new_path)
            ))
```
And when taking an element from the stack and it's an iterator, then get its next element. The loop in the bottom-up case could then be replaced with just `stack.append(walk_dirs)`.

**Comment by Ovsyanka83** [comment]
Honestly, I am quite baffled by your discovery. I agree that changing even small implementation details is dangerous on such an old function. However, I also would never think about using `walk` this way. I will leave the decision to the *adults* here :))

I do not think that this affects our changes for `pathlib.Path.walk` though because no old code is going to be broken.

Thank you for finding this out and notifying us!

P.S. If we do decide to agree with you on this and implement "late dirnames modifications", then we must also add this as a test case to prevent any breaking changes in the future.

**Comment by jonburdo** [comment]
It might have resulted in cleaner, more predictable behavior if `os.walk` had always been designed to copy `dirs` right before beginning to iterate over it, preventing this late editing. Editing lists while iterating over them is generally discouraged for good reason. Note that in [this example](https://github.com/python/cpython/issues/89727#issuecomment-1369298204) with recursive `os.walk`, you get the same results if you change `del demo[1][1]` to `del demo[1][0]`.

But this has existed for a while, and people may rely on it, so it would probably be ideal to maintain the old behavior.

I looked into this a bit and your suggestions @pochmann :
> ```
>             stack.append((
>                 new_path
>                 for dirname in dirs
>                 for new_path in [join(top, dirname)]
>                 if followlinks or not islink(new_path)
>             ))
> ```
Something like this should work, although note that `top` will be changed between calls to `next` on this generator, which causes problems. We need the same value that `top` is set to when this generator is created. Something like this could work although it's pretty ugly:
```python
            stack.append((
                new_path
                for top, dirs in [(top, dirs)]
                for dirname in dirs
                for new_path in [join(top, dirname)]
                if followlinks or not islink(new_path)
            ))
```
It might also be better to just do `stack.append((top, dirs))` and handle the joining and link logic when these values are used from the stack.

**Comment by pochmann** [comment]
> note that `top` will be changed between calls to `next` on this generator, [...] Something like this could work although it's pretty ugly

Another option, with itertools (untested, having computer problems):
```python
            new_paths = map(join, repeat(top), dirs)
            stack.append(
                new_paths if followlinks else
                filterfalse(islink, new_paths)
            )
```
I see you have bigger changes in mind in your PR. Will look later.

**Comment by jonburdo** [comment]
Is there any consensus on this issue of whether or not to support this old behavior of supporting late editing of dirs?

**Comment by Ovsyanka83** [comment]
I would vote against it. But I do not think there is a clear consensus. We do not have any proof for or against the hypothesis that someone uses this behavior. And even if someone does, it feels more like a bug that we have not fixed for years.

**Comment by jonburdo** [comment]
> And even if someone does, it feels more like a bug that we have not fixed for years.

Yeah, I think I'd agree. I'll leave the PR to support it ( https://github.com/python/cpython/pull/100703 ) open for now.

I also converted to a draft my PR for an iterative `os.fwalk`, since this will introduce the same situation.

**Comment by barneygale** [comment]
Thanks for your contributions, everyone. I've logged the late edits issue separately:

- #102932

Is there anything left to do here?

**Comment by jonburdo** [comment]
> Is there anything left to do here?

`shutil.rmtree` is still recursive. I was thinking of giving that a shot when I have time.

I also started a PR for `os.fwalk` linked here, but marked it as a draft, since I wasn't sure what the decision on handling late-editing would be. I'm happy to revisit it if it seems worth. That would at least make `os.fwalk` consistent with `os.walk`.

Should those be tracked in separate issues?

**Comment by barneygale** [comment]
> `shutil.rmtree` is still recursive. I was thinking of giving that a shot when I have time.
> 
> I also started a PR for `os.fwalk` linked here, but marked it as a draft, since I wasn't sure what the decision on handling late edits would be. I'm happy to revisit it if it seems worthwhile. That would at least make `os.fwalk` consistent with `os.walk`.
> 
> 

I think it's sensible to make `fwalk()` consistent with `walk()`, even if it's affected by #102932. Do you think we could then implement `rmtree()` atop `walk()` / `fwalk()`?

Both problems should be tracked here, I think, because they directly relate to the original report.

**Comment by jonburdo** [comment]
Okay, makes sense. I'll revisit that `os.fwalk` PR.

> Do you think we could then implement `rmtree()` atop `walk()` / `fwalk()`?

I'm not sure right off. I'll look into it.

**Comment by jonburdo** [comment]
Following up on implementing `rmtree` with `walk` / `fwalk`, there are some `os.DirEntry` calls, `entry.is_junction()` and `entry.is_symlink()`. These probably could be done with `os.path` functions, but I'm not sure how much of a difference this would make. It still might just be more straightforward either way to have `rmtree` use its own stack rather than using `walk` / `fwalk`.

It seems like a `walk` function that returns the `os.DirEntry` objects in the dirs list would be nice - both for this case and maybe as a user-facing function. Even if the status of `entry.stat()` cache is unknown, it would still be useful. Any thoughts? See related topic: https://discuss.python.org/t/get-direntry-objects-collected-during-os-walk/8143

**Comment by barneygale** [comment]
> Following up on implementing `rmtree` with `walk` / `fwalk`, there are some `os.DirEntry` calls, `entry.is_junction()` and `entry.is_symlink()`. These probably could be done with `os.path` functions, but I'm not sure how much of a difference this would make.

I think the `entry.is_symlink()` call can be replaced with `os.path.islink()` quite readily. It's there's to address a timing issue, and arguably shouldn't rely on the `entry` at all.

The `entry.is_junction()` call is interesting. Perhaps `walk()` should treat junctions like symlinks, which seems to be the behaviour in some other `os` functions. Alternatively we could add a *followjunctions* argument to `os.walk()` (and `fwalk()`).

I hereby summon @eryksun for his Windows filesystem expertise!

> It seems like a `walk` function that returns the `os.DirEntry` objects in the dirs list would be nice - both for this case and maybe as a user-facing function.

Maybe. There's a really high threshold to meet for additions to CPython, and I'd like to see if we can make `walk()` work first.

**Comment by barneygale** [comment]
Going deeper, should `os.DirEntry.is_dir(follow_symlinks=False)` return `False` for junctions, as it would for symlinks to directories?

**Comment by eryksun** [comment]
POSIX has no file type that corresponds to a Windows filesystem junction, which is a name-surrogate reparse point that's tagged `IO_REPARSE_TAG_MOUNT_POINT`. POSIX implements mountpoints differently, using a centralized table, whereas Windows uses a hybrid system that's partly decentralized using reparse points. Python generally handles junctions as directories, such as when copying (e.g. `shutil.copytree()`). However, junctions are special cased to be handled like symlinks when deleting (e.g. `os.unlink()` and `shutil.rmtree()`). Windows shells such as CMD and Explorer handle junctions similarly.

`os.unlink()` special cases directory symlinks and junctions by calling WinAPI `RemoveDirectoryW()`. The latter has special handling when deleting a junction that targets the root path of a volume GUID device name (e.g. "\\??\\Volume{12345678-1234-1234-1234-123456781234}\\", i.e. a volume mount point). In this case, it calls `DeleteVolumeMountPointW()`, which attempts to remove the mountpoint from the system's mountpoint database. The latter is the basis for the canonical DOS path of a volume, as returned by `GetFinalPathNameByHandleW()`, so it's important to keep it accurate. (If the caller lacks administrator access, updating the mountpoint database will fail, but the directory might still get deleted. To avoid this, volume mountpoints should be secured to only allow deletion by an administrator.)
  
In some respects, a junction is similar to a bind mountpoint on POSIX. Of note here is that the path parsing behaviors of mountpoints and symlinks are distinctly different. For example, when opening "C:\\Mount\\mountpoint\\spamlink", the system uses "C:\\Mount\\mountpoint" as the opened path when resolving "spamlink". If, for example, "spamlink" targets "..\\spam", the system will resolve it as "C:\\Mount\\spam". In other words, "mountpoint" is handled like a POSIX directory mount point, not like a symlink. This is similar to how "/mnt/mountpoint/spamlink" would be resolved on POSIX as "/mnt/spam". On the other hand, when opening "C:\\Links\\symlink\\spamlink", the resolved path of "spamlink" will depend on the resolved path of "symlink", which could be a path on another device or even on another machine.

The other distinct difference with junctions is that they're always resolved on the machine where they reside. The system enforces this by requiring the target of a junction to be a directory on a local filesystem. Say, for example, you have a local directory that's shared as "\\\\server\\share" and you want to link a local directory "F:\\spam" as "\\\\server\\share\\spam". You can easily implement this using a junction since reparsing the junction is handled locally on your server. Now say that a client user makes a local copy of "\\\\server\\share". The client needs to copy the junction "\\\\server\\share\\spam" as a regular directory. It shouldn't be copied as a reparse point. At best that would be meaningless on the client system, and at worst it would be dangerously wrong if the client coincidentally has a local "F:\\spam" directory. 

Symlinks, on the other hand, are always resolved on the client machine. Thus if you tried to implement "\\\\server\\share\\spam" as a symlink to "F:\\spam", the client would attempt to resolve the symlink as a path on the client's *local drive* "F:". By default the system won't even allow this (thankfully) because the default policy is to forbid evaluating remote-to-local symlinks. To avoid the dangerous nonsense of a remote-to-local symlink, you could share "F:\\spam" as "\\\\server\\spam_share" and link to that. This is sensible, but it's probably disallowed. The default symlink policy forbids evaluating remote-to-remote symlinks. (Local-to-local and local-to-remote symlinks are allowed by default.)

**Comment by barneygale** [comment]
Thank you Eryk, that's very useful to know.

Some possible ways forward:

1. We could add a *followjunctions* argument to `os.walk()`, defaulting to `True`. We could then re-implement `rmtree()` atop `walk()`, setting *followjunctions* to `False`
2. We could add a more generic `walk()` API that exposes the `os.DirEntry` objects, and then re-implement `rmtree()` atop that.
3. We could rewrite `rmtree()` to be iterative, but without making it use `walk()`.

I'm keen to avoid (3) if we can. The iterative versions of these functions *are* more complicated than their recursive predecessors, and so I'd rather have a small number of high-quality implementations.

**Comment by eryksun** [comment]
Adding `followjunctions=True` to `os.walk()` works for me. By default, junctions should be traversed the same as directories, such as in `shutil.copytree()`. OTOH, deleting a tree would use `followjunctions=False`, such as in `shutil.rmtree()`.

**Comment by jonburdo** [comment]
> Adding `followjunctions=True` to `os.walk()` works for me

I'll give this a shot.

**Comment by barneygale** [comment]
I spotted a problem with using `os.walk()` / `os.fwalk()` to implement `rmtree()`: for historical reasons, `walk()` places symlinks to directories in the `dirnames` list _even if_ `followlinks` is set to `False`. 

This is a wart we removed when we added `pathlib.Path.walk()`. So there's a possible alternative here:

1. Add a `fwalk()` method to `pathlib.Path`
2. Add a `follow_junctions` argument to `pathlib.Path.walk()`
3. Implement `shutil.rmtree()` atop those methods.

I'm going to have a crack at this.

**Comment by jonburdo** [comment]
Won't this be significantly slower? On my system, with a tree with ~5000 directories under it, `Path.walk` takes about 20-30% longer than `os.walk`. `shutil.rmtree` takes a lot longer of course, but even a 5% slowdown seems significant.

Why not adjust `os.walk` to allow returning `os.DirEntry` objects? This could even be an internal function for now.

**Comment by eryksun** [comment]
> `os.walk()` places symlinks to directories in the `dirnames` list even if `followlinks` is set to `False`.

Preferably `dirnames` would be ignored when implementing `rmtree()` using a bottom-up walk. The directories in that list should have been removed already. To get the desired behavior regarding symlinks/junctions, a parameter could be added to `os.walk()` to enable yielding non-followed symlinks/junctions in `filenames` instead of `dirnames`.

**Comment by barneygale** [comment]
> Won't this be significantly slower? On my system, with a tree with ~5000 directories under it, Path.walk takes about 20-30% longer than `os.walk`. `shutil.rmtree` takes a lot longer of course, but even a 5% slowdown seems significant.

Yep good point, it might be too slow. I also measure `Path.walk()` as ~20% slower than `os.walk`, though my implementation of `Path.fwalk()` is within 5% of `os.fwalk()`.

OTOH, `Path.walk()` treats symlinks exactly as required by `rmtree()`, unlike `os.walk()`.

> Why not adjust `os.walk` to allow returning `os.DirEntry` objects? This could even be an internal function for now.

What does the API look like? 

> Preferably `dirnames` would be ignored when implementing `rmtree()` using a bottom-up walk. The directories in that list should have been removed already.

In both `os.walk()` and `Path.walk()`, if an `os.DirEntry.is_dir()` call succeeds but a subsequence `scandir()` of the same directory fails, then the directory will appear in `dirnames` of its parent but _not_ be visited afterward. If `rmtree()` is then implemented on top, this behaviour causes test failures in `test_tempfile`.

Arguably `Path.walk()` could yield a result with empty `dirnames` and `filenames` when `scandir()` fails.

> To get the desired behavior regarding symlinks/junctions, a parameter could be added to `os.walk()` to enable yielding non-followed symlinks/junctions in `filenames` instead of `dirnames`.

`follow_symlinks_for_realsies`? :)

**Comment by barneygale** [comment]
> Preferably `dirnames` would be ignored when implementing `rmtree()` using a bottom-up walk. The directories in that list should have been removed already.

In both `os.walk()` and `Path.walk()`, if an `os.DirEntry.is_dir()` call succeeds but a subsequence `scandir()` of the same directory fails, then the directory will appear in `dirnames` of its parent but _not_ be visited afterward. If `rmtree()` is then implemented on top, this behaviour causes test failures in `test_tempfile`.

Arguably `Path.walk()` could yield a result with empty `dirnames` and `filenames` when `scandir()` fails.

> To get the desired behavior regarding symlinks/junctions, a parameter could be added to `os.walk()` to enable yielding non-followed symlinks/junctions in `filenames` instead of `dirnames`.

`follow_symlinks_for_realsies`? :)

**Comment by eryksun** [comment]
`rmtree()` would use a bottom-up walk instead of the default top-down walk. Thus if `os.scandir()` fails for a directory, it does so before `dirnames` gets emitted for the parent directory. Each directory would be deleted via `os.rmdir(dirpath)`, after first removing all of the yielded `filenames`. At this point, subdirectories have already been visited and removed.

> follow_symlinks_for_realsies? :)

I was being serious. Maybe `strict_dirnames=False`.

**Comment by barneygale** [comment]
> `rmtree()` would use a bottom-up walk instead of the default top-down walk. Thus if `os.scandir()` fails for a directory, it does so before `dirnames` gets emitted for the parent directory. Each directory would be deleted via `os.rmdir(dirpath)`, after first removing all of the yielded `filenames`. At this point, subdirectories have already been visited and removed, and `dirnames` would be completely ignored.

Even with `topdown=False`, the function _internally_ visits directories from top to bottom (there's no other way), and what I wrote about there being no 3-tuple yielded for un-`scandir()`able directories still applies.

> I was being serious. Maybe `strict_dirnames=False`.

Sorry, I didn't mean to imply you weren't being serious, only that I thought the naming wasn't obvious. `strict_dirnames` is OK but I'm generally not a fan of complicating a function signature to make things behave how they should always have behaved. And I don't really see the need to add this when we already have `pathlib.Path.walk()`, which gets symlink handling right -- in fact, I believe this improvement was one of the things Brett Cannon thought tipped the balance in favour of adding `Path.walk()` at all.

**Comment by eryksun** [comment]
> what I wrote about there being no 3-tuple yielded for un-`scandir()`able directories still applies.

Say `os.walk()` calls `scandir('d1')`, which contains subdirectory "d2" and no files. It pushes `('d1', ['d2'], [])` on the stack (a list), and then it pushes "d2" on the stack. The next iteration pops "d2" from the stack and calls `os.scandir('d1/d2')`. If we assume it fails, then the default `onerror` handler of `rmtree()` re-raises the exception, and `rmtree()` fails, as it should. If `os.scandir('d1/d2')` succeeds, and "d2" contains file "f", then `('d1/d2', [], ['f'])` gets pushed on the stack. The latter gets yielded to `rmtree()`, which calls `os.unlink('d1/d2/f')` and then `os.rmdir('d1/d2')`. Finally, `('d1', ['d2'], [])` gets yielded to `rmtree()`, which calls `os.rmdir('d1')`, and we're done.

**Comment by barneygale** [comment]
You're totally right, of course Eryk :-). I must have made a mistake when I tried that approach previously, as when I try it now I get zero test failures from `test_tempfile`.

**Comment by jonburdo** [comment]
> I also measure `Path.walk()` as ~20% slower than `os.walk`, though my implementation of `Path.fwalk()` is within 5% of `os.fwalk()`.

Maybe that is close enough to just use `Path.fwalk()`, although it's probably worth directly comparing any refactor of `rmtree` to the existing implementation to make sure.
 
> > Why not adjust `os.walk` to allow returning `os.DirEntry` objects? This could even be an internal function for now.
> 
> What does the API look like?

It could simply mean introducing a kwarg like `direntry=False`. This seems questionable to introduce a kwarg on `os.walk` itself, but there are a couple of options:
1. Introduce `os._walk` with `direntry=True`, so this isn't public-facing. Maybe `os.walk` just calls `os._walk` always with `direntry=False`.
2. More generally, implement whatever walk function would be most useful for `rmtree` and whatever else should make use of it (`copytree` maybe) as something like `os._walk` or `os._fwalk`. This might essentially be what you've already done with `Path.fwalk`. On the other hand, implementing a purely bottom up function that only returns `os.DirEntry` objects for dirs could end up being simpler and faster by enough to be worthwhile.

Either of these options could also help avoid introducing `strict_dirnames=False`. And whether it's used for this or not, I think `Path.fwalk` is good to have.

Personally I don't see an ideal solution, which is also comfortably made to be public-facing. Even if `pathlib`'s walk function were as fast as `os`'s , in some cases I specifically want string paths and avoiding `Path` creation altogether is preferable. I'd personally love to have a public facing `os.walk2` or something which implements `os.walk` as we'd do it now if backwards compatibility weren't an issue. This is probably more extreme, ugly, confusing, etc than simply adding kwargs to `os.walk` though.

**Comment by barneygale** [comment]
What sort of thing would `os._walk(direntry=True)` yield? A 2-tuple of `(path, entries)`? A 3-tuple of `(path, direntries, fileentries)`? A 4-tuple of `(path, dirnames, direntries, filenames)`? Bearing in mind we want users to be able to influence the walk by adjusting the directory list, _and_ distinguish between between files/directories/symlinks for the purposes of walking.

**Comment by eryksun** [comment]
> public facing `os.walk2` or something which implements `os.walk` as we'd do it now

If new functions are added, I suggest `os.path.walk()` and `os.path.fwalk()`. It's where I think they belong, in correspondence with `pathlib.Path`. I'd prefer if it yielded `os.DirEntry` instances for `dirpath` and for the items in `dirnames` and `filenames`. There's no point in throwing away the basic stat information. I'd prefer for the `onerror` handler to match `shutil.rmtree()`, with three parameters: `function`, `path`, and `excinfo` (or just the exception instance instead of `excinfo`).

**Comment by jonburdo** [comment]
> What sort of thing would `os._walk(direntry=True)` yield?

It would simply replace members of `dirnames` and `filenames` with `os.DirEntry` objects. So this would basically be the current `os.walk` implementation with a two line change, replacing `dirs.append(entry.name)` with `dirs.append(entry if direntry else entry.name)` - same for `nondirs`. And the same could be done for `fwalk`.

Of course if we just want a separate implementation, public or not, then we could maybe just get rid of `direntry` and `strict_dirnames` kwargs altogether. Instead we handle symlinks properly and always return `os.DirEntry` objects (maybe still a string path for `dirpath` though).

**Comment by barneygale** [comment]
I'd be interested to see an implementation! You'd need to support `os.walk()`'s behaviour of adding symlinks to directories to `dirnames` but not walking into them, and also `rmtree()`'s requirement that those symlinks appear in `filenames` for `unlink()`ing.

For now I'm -1 on adding new _public_ functions. I wrestled with the `os.walk()` API [when we added `pathlib.Path.walk()`](https://discuss.python.org/t/add-pathlib-path-walk-method/12968). I was initially convinced that the API could be significantly improved, but the more I worked it, the more I realised that `(path, dirnames, filenames)` covers 99% of use cases most elegantly.

**Comment by eryksun** [comment]
> It would simply replace members of `dirnames` and `filenames` with `os.DirEntry` objects. So this would basically be the current `os.walk` implementation with a two line change, replacing `dirs.append(entry.name)` with `dirs.append(entry if direntry else entry.name)` - same for `nondirs`. And the same could be done for `fwalk`.

Returning `os.DirEntry` objects doesn't avoid the need to fix the problem with `dirnames`. Without `strict_dirnames`, `rmtree()` would have to try to delete every item in `dirnames`, of which everything except unfollowed symlinks and junctions was already deleted. OTOH, with `strict_dirnames`, `rmtree()` can ignore `dirnames`. 

> Of course if we just want a separate implementation, public or not, then we could maybe just get rid of `direntry` and `strict_dirnames` kwargs altogether. Instead we handle symlinks properly and always return `os.DirEntry` objects (maybe still a string path for `dirpath` though).

The split between `dirnames` and `filenames` is convenient. `rmtree()` would simply loop over `filenames` to `unlink()` each item, and ignore `dirnames`.

I'd prefer to return `dirpath` as a `DirEntry`. It has basic stat info that's very cheap on Windows (size, timestamps, file attributes, reparse tag) -- much cheaper than `os.[l]stat()` -- and it supports `__fspath__()` for use with `os`, `os.path`, and `pathlib.Path`.

**Comment by jonburdo** [comment]
I think my suggestion wasn't clear enough - I also might be missing something.

> Returning os.DirEntry objects doesn't avoid the need to fix the problem with dirnames.

Right, I'm discussing these as separate matters. Imagine that we take these steps:
1. Copy the `os.walk` implementation to `os._walk`
2. Add `direntry=False`, which if set to `True` returns `os.DirEntry` objects instead of strings.
3. Add `strict_dirnames=False`.
4. Add `followjunctions=True`.
We can create an `os._fwalk` function similarly. Now we have private functions for things like `rmtree`. Then we realize we don't need `strict_dirnames=False` on this function, so we remove `strict_dirnames`, but keep the functionality of `strict_dirnames=True`. Maybe we also realize that we don't need `direntry=False` either, so we remove `direntry`, keeping the functionality of `direntry=True`.

Maybe a simpler way to think about this: We want the behavior of `Path.walk` and @barneygale's `Path.fwalk`, but without any `pathlib` objects involved. Instead of `Path` objects we always return `os.DirEntry` objects. The benefit here is speed and access to the `os.DirEntry` objects.

I didn't mean to suggest that we remove the split between `dirnames` and `filenames`.

> I'd prefer to return dirpath as a DirEntry

You might be right.

Anyways, I need to think a little more about it and maybe I can get an implementation done soon to demonstrate.

**Comment by barneygale** [comment]
My view (open to change) is that we don't need a `strict_dirnames` argument or a new implementation of `walk()` that supports this behaviour when we already have `pathlib.Path.walk()`: a method that exists in part _because_ it implements the `strict_dirnames` behaviour that we need for `rmtree()`.

I have a WIP implementation using pathlib in #103164

The only downside I can see is performance, and we haven't yet put a number to it. I'll try to do that soon. I think only Windows will be affected, if that. (Is it possible we'll make `rmtree()` slightly more reliable on Windows by pausing for longer between deletions? :joy:). And there are still avenues available to us to improve pathlib performance that could make up the difference.

That said, I'm an ardent pathlib fanboy, so it would be good to get other opinions on this.

**Comment by eryksun** [comment]
> Is it possible we'll make `rmtree()` slightly more reliable on Windows by pausing for longer between deletions?

Microsoft has improved this for the cases in which a file or directory is open with delete sharing (e.g. watching a directory for changes, or a malware scanner). In Windows 10, `DeleteFileW()` uses a POSIX delete if supported by the filesystem. In Windows 11, `RemoveDirectoryW()` uses a POSIX delete if the filesystem supports it. For NTFS, a POSIX delete renames the file or directory into a hidden system directory named "\\$Extend\\$Deleted". For example:

```pycon
>>> os.mkdir('spam')
>>> h = win32file.CreateFile('spam', 0x0001_0000, 7, None, 3, 0x0200_0000, None)
>>> os.rmdir('spam')
>>> win32file.GetFinalPathNameByHandle(h, 0)
'\\\\?\\C:\\$Extend\\$Deleted\\000B000000041FC810B3AEBC'
```

OTOH, if a file or directory is opened without delete sharing (e.g. via Python's `open()` or `os.chdir()`), then trying to delete it will fail as a sharing violation (error 32). Also, if a file is mapped as an executable image (e.g. a process executable or loaded DLL), then trying to delete it will fail with access denied (error 5). There's nothing we can do about this in general, short of terminating the offending processes. We can use an error handler that retries the delete with an increasing timeout up to maximum total wait time, which is useful for a race condition between closing and deleting a file.

In issue #101601, I proposed adding job object support to `subprocess.Popen` and better support for sending `SIGBREAK` to a console process group. This would support cleanly and reliably ensuring that a process tree has terminated before attempting to remove files and directories in use by the processes.

**Comment by jonburdo** [comment]
Here's a rough draft of walk/fwalk functions that return `os.DirEntry` objects: https://github.com/python/cpython/pull/103234

It needs a couple changes, including the `follow_junctions` kwarg if we'd want that. It's pretty similar to the current `Path.walk` and to my `os.fwalk` function in https://github.com/python/cpython/pull/100347

I think your `rmtree` implementation and pathlib functions in https://github.com/python/cpython/pull/103164 look good. I'd honestly love to have public walk functions returning `os.DirEntry` objects and also `Path.fwalk`. Of course for now though, we can use whatever private functions seem most reasonable for `rmtree`.

**Comment by barneygale** [comment]
I've raised this on discourse: https://discuss.python.org/t/using-iterative-filesystem-walk-instead-of-recursive/21955/21

**Comment by barneygale** [comment]
A few updates:

- `glob.glob()` will be fixed by #116392
- `glob.iglob()` likewise
- `os.fwalk()` will be fixed by #119638
- `shutil.rmtree()` (non-fd version) will be fixed by #119634
- `shutil.rmtree()` (fd version) will be fixed by an upcoming PR

**Comment by barneygale** [comment]
One more for the pile: `shutil.copytree()` also calls itself recursively.
